### PR TITLE
[s04a] Logistic regression

### PR DESCRIPTION
**Pull request by colinhex**
Hi everyone,

I expanded on the outline of the original script and added more thorough recaps & summary.

I'm happy to make further changes on request.

### CODE REVIEW

File: logistic-regression.md [hunk 1]
```diff
@@ -0,0 +1,752 @@
...
+    -2\mathbf{X}^T(\mathbf{y}-\mathbf{X}\mathbf{w})
+$$
+$$
+    -2(\mathbf{X}^T\mathbf{y}-\mathbf{X}^T\mathbf{X}\mathbf{w}^*)=0
+$$
+$$
+    \mathbf{X}^T\mathbf{y}=\mathbf{X}^T\mathbf{X}\mathbf{w}^*
+$$
+$$
+    \frac{\mathbf{X}^T\mathbf{y}}{\mathbf{X}^T\mathbf{X}}=\mathbf{w}^*
```
Comment (vinith2) [inline logistic-regression.md:56]: You cannot divide a  vector by a matrix. You can write an inverse, then you need to assume that inverse exists. Pseudo inverse is more appropriate here.

Comment (colinhex) [reply logistic-regression.md]: Changed it to an inverse, where it came up.

File: logistic-regression.md [hunk 2]
```diff
@@ -0,0 +1,752 @@
...
+Think of two images, one is white noise and one is of a banana. Which of these pictures carries more information? You might think
+the banana, since we recognize it, and it conjures up relations in our minds. White noise on the other hand is meaningless to us.
+
+![alt](./images/white-noise.png) ![alt](./images/Banana-Single-2.jpg)
+
+Say you want to send this data to a friend. Can you compress one more than the other? There is no way to relate any pixel value
+to another in a random image, so you indeed have to send all the data if your friend should receive it without loss. The noise is
+a state of maximum information or *maximum entropy*.
+
+When we speak of the **cross-entropy** between two probability distributions over the same set of events we are talking about
```
Comment (vinith2) [inline logistic-regression.md:392]: May be add the equation of the cross entropy between two probability distributions

File: logistic-regression.md [hunk 3]
```diff
@@ -0,0 +1,752 @@
...
+
+...where $\mu_{ic} = [softmax(w^T_1x_i,..., w^T_Cx_i)]_c$ and $y_{ic} = \textbf{1}(y_i = c)$. The right hand
+side of this simplification [can be read](https://en.wikipedia.org/wiki/Cross_entropy#Relation_to_maximum_likelihood) as the product of the estimated probabilities for classes $c$ to the
+power of their occurrences in the predictions, which is to say their actual probability. And if we substitute these definitions in the previous equations we get a more familiar equation.
+    
+$$                                                              
+    \sum_{i=1}^k ln\left(\prod_{c=1}^C \mu^{y_{ic}}_{ic}\right) = \sum_{i=1}^k \sum_{c=1}^C ln(\mu^{y_{ic}}_{ic}) = \sum_{i=1}^k \sum_{c=1}^C y_{ic} ln(\mu_{ic})
+$$
+
+The term $-\sum_{c=1}^C y_{ic} ln(\mu_{ic})$ is also defined as **cross-entropy**. This means that
```
Comment (vinith2) [inline logistic-regression.md:410]: Can you show why this is cross entropy?  As cross entropy is between the two probability distribution. How is the term you mentioned related to the two probability distributions?

File: logistic-regression.md [hunk 4]
```diff
@@ -0,0 +1,752 @@
...
+
+For our two classes, we know $p(y_i = -1) = 1 - p(y_i = 1)$ and this simplifies to:
+
+$$
+    NNL(\mathbf{w}) = -\sum_{i=1}^k \left(y_{i1} ln(\mu_{i1}) + y_{i2} ln(\mu_{i2})\right) =  -\sum_{i=1}^k \left( y_i ln(\mu_i) + (1-y_i) ln(1-\mu_i)\right)
+$$
+
+Another way of picturing this equation is as measuring the cross-entropy between two probability distributions
+$(y_i, (1-y_i))$ and $(\mu_i, (1-\mu_i))$, as in how close the distribution of the predictions is to
+the desired, actual distribution of the data. If we examine $\mu_i$ and $(1 - \mu_i)$ we find that they simplify considerably:
```
Comment (vinith2) [inline logistic-regression.md:441]: shouldn't this be : "desired/ actual  or desired (actual)"

File: logistic-regression.md [hunk 5]
```diff
@@ -0,0 +1,752 @@
...
+    NLL(\mathbf{w}) = -\sum_{i=1}^k \left(y_i ln\left(\frac{1}{1 + e^{w^Tx}}\right) + (1-y_i) ln\left(\frac{1}{1 + e^{-w^Tx}}\right)\right)
+$$
+$$
+    = \sum_{i=1}^k \left(y_i ln(1 + e^{w^Tx}) + (1-y_i) ln(1 + e^{-w^Tx})\right)
+$$
+
+If we further consider the function for binary cases \{-1, 1\}, then we can simply write: 
+
+$$
+    NLL(\mathbf{w}) = \sum_{i=1}^k ln(1 + e^{-y_iw^Tx_i}) \quad y_i \in \{-1, 1\}
```
Comment (vinith2) [inline logistic-regression.md:466]: Elaborate a bit here. As how do we get this equation.

File: logistic-regression.md [hunk 6]
```diff
@@ -0,0 +1,752 @@
...
+    \lim_{\hat{y} \rightarrow +\infty} ln(1 + e^{-y\hat{y}}) = \begin{cases} +\infty &\quad y = -1 \\ 0 &\quad y = 1 \end{cases}
+$$
+
+The same applies to the other direction, the loss goes to zero if the prediction matches the outcome, otherwise it grows to infinity.
+
+$$
+    \lim_{\hat{y} \rightarrow -\infty} ln(1 + e^{-y\hat{y}}) = \begin{cases} 0 &\quad y = -1 \\ +\infty &\quad y = 1 \end{cases}
+$$
+
+If you can still remember, we derived this loss to get rid of the least squares loss, which we
```
Comment (vinith2) [inline logistic-regression.md:485]: Instead of "If you can still remember", point to the part of the text where you derived this.

File: logistic-regression.md [hunk 7]
```diff
@@ -0,0 +1,752 @@
...
+
+y = numpy.linspace(-1.5, 3.5, 1000)
+y_hat = 1  # Always predict 1.
+
+fig, ax = plt.subplots(1, 1, figsize=(10, 4))
+line1 = ax.plot(y, loss_l2(y, y_hat), '--r', lw=1, label='L2-Loss')
+line2 = ax.plot(y, loss_01(y, y_hat), '-r', lw=1, label='L01-Loss')
+line3 = ax.plot(y, loss(y, y_hat), '-b', lw=1, label='LL-Loss')
+ax.legend(('L2-Loss', 'L01-Loss', 'Logistics-Loss'))
+plt.setp(ax, xlabel='y', ylabel='Loss for y_hat = 1', ylim=(-0.1, 2.0), xlim=(-1, 2.5))
```
Comment (vinith2) [inline logistic-regression.md:512]: Looks like,logistic loss is also different compared  to the L_{01} loss. Can you write why this logistic loss is better  compared to L_2 loss?

File: logistic-regression.md [hunk 8]
```diff
@@ -0,0 +1,752 @@
...
+the partial derivatives to zero and obtain our optimal weights $\mathbf{w}$.
+
+$$
+    \nabla_{\mathbf{w}}NLL(\mathbf{w}) = \nabla_{\mathbf{w}}\sum_{i = 1}^{n}(\mu_i-y_i)x_i = \nabla_{\mathbf{w}}\mathbf{X}^T(\mu-y)
+$$
+$$
+    = \mathbf{X}^T\left(\nabla_{\mathbf{w}}\frac{e^{w^T_cx}}{\sum_{c=1}^{C}e^{w^T_cx}}\right) 
+$$
+
+Now there are some good news and bad news. The good news is, we won't have to calculate this derivative and set it to
```
Comment (vinith2) [inline logistic-regression.md:530]: Why is it a  good news that we cant set it to 0?

File: logistic-regression.md [hunk 9]
```diff
@@ -0,0 +1,752 @@
...
+zero in order to find the optimal weights. The bad news is, that we can not do so and have to find another way. Since the weights are tugged away in the exponents, here, we will not 
+find a closed form expression for logistic regression like we did when deriving the linear least squares loss.
+
+### Gradient (Steepest) Descend
+
+Using gradient descend we can approximate local minima or maxima of functions in any number of dimensions. We iteratively
+approach the extrema by moving along the direction of the gradient of a function. You know the derivative of a function gives us
+a value for its steepness at every point in space. In gradient (steepest) descend, we find the **direction**, you 
+guessed it, of the **steepest descend** and move a distance with a certain factor, the **step size**.
```
Comment (vinith2) [inline logistic-regression.md:539]: Maybe a bit less conversation text, would be easy to follow.  
Can you write the relation with direction of steepest descent to that of gradient. Some introductory materials on gradient descent will have this reason.

File: logistic-regression.md [hunk 10]
```diff
@@ -0,0 +1,752 @@
...
+### Gradient (Steepest) Descend
+
+Using gradient descend we can approximate local minima or maxima of functions in any number of dimensions. We iteratively
+approach the extrema by moving along the direction of the gradient of a function. You know the derivative of a function gives us
+a value for its steepness at every point in space. In gradient (steepest) descend, we find the **direction**, you 
+guessed it, of the **steepest descend** and move a distance with a certain factor, the **step size**. 
+If we start with some initial guess and repeat this process a certain number of times, where we end up must be close
+to a **local minimum**. We can formulate this in an algorithm as follows:
+
+$\textit{for} \quad t \in \{1, ..., t_{max}\} \quad \textit{do}:$
```
Comment (vinith2) [inline logistic-regression.md:543]: Put it inside a latex algorithm environment.

File: logistic-regression.md [hunk 11]
```diff
@@ -0,0 +1,752 @@
...
+```{code-cell} ipython3 
+:tags: [hide-input]
+x_ = numpy.linspace(0, 2, 200)
+y_ = numpy.linspace(-0.5, 3, 200)
+xx, yy = numpy.meshgrid(x_, y_)
+zz = 0.5*(xx**2 - yy)**2 + 0.5*(xx - 1)**2
+
+fig = plt.figure(figsize=(7,7))
+ax = plt.axes(projection='3d')
+ax.plot_surface(xx, yy, zz, cmap=cm.coolwarm,  rcount=100, ccount=100);
```
Comment (vinith2) [inline logistic-regression.md:569]: what does the color signify?  Is it the z-axis. Maybe add a colorbar as well?

File: logistic-regression.md [hunk 12]
```diff
@@ -0,0 +1,752 @@
...
+### Step Size and Learning Rate
+
+As stated, now visible, the weights will be updated iteratively to approach the local minimum in direction
+of the steepest descend. If the gradient will be near zero in an area around the true minimum, we will approach
+it more slowly. One can imagine the impact of this optimization on various surfaces and how initial guess, step size and the number
+of iterations play into finding a close-to-optimal minimum.
+
+If gradient descend is used as an optimization technique for a model, then the step size can be seen, among else, as its 
+**learning rate**. If we set it too high, we might overshoot, oscillate or even diverge with our weights. If we set it
+to low, we might need a *large* number of iterations to find it.
```
Comment (vinith2) [inline logistic-regression.md:688]: Can you plot the gradient descent with different step size? Essentially showing the oscillations or diverge phenomena you  are talking about.

File: logistic-regression.md [hunk 13]
```diff
@@ -0,0 +1,752 @@
...
+
+$$                                                                   
+    NLL_{\mathcal{l}2}(\mathbf{w}) = \frac{1}{n}\sum_{i=1}^k (1 - y_iw^Tx_i)^2
+$$
+
+Finding the optimal parameters by taking the derivative is not possible since there does not exist a closed
+form expression for our problem.
+
+$$
+\mathbf{X}^T(\nabla_{\mathbf{w}}\frac{e^{w^T_cx}}{\sum_{c=1}^{C}e^{w^T_cx}})
```
Comment (vinith2) [inline logistic-regression.md:728]: Should this be equated to 0? I don't understand what this term signify.

File: logistic-regression.md [hunk 14]
```diff
@@ -0,0 +1,752 @@
...
+$$
+
+If we double down on and work to simplify this expression we can establish a relationship to another concept in information theory.
+
+### Cross-Entropy
+
+Think of two images, one is white noise and one is of a banana. Which of these pictures carries more information? You might think
+the banana, since we recognize it, and it conjures up relations in our minds. White noise on the other hand is meaningless to us.
+
+![alt](./images/white-noise.png) ![alt](./images/Banana-Single-2.jpg)
```
Comment (vinith2) [inline logistic-regression.md:386]: Can you center this image?

**Review by vinith2 — COMMENTED** [review]
Hi, the page looks good. I have added some for some specific lines in the script.

File: logistic-regression.md (hunk 8 above)
Comment (colinhex) [reply logistic-regression.md:530]: It specifically says it is bad news that we can't set it to zero.

File: logistic-regression.md (hunk 9 above)
Comment (colinhex) [reply logistic-regression.md:539]: I'm sorry can you rephrase this comment? I have a hard time understanding it.

File: logistic-regression.md (hunk 14 above)
Comment (colinhex) [reply logistic-regression.md:386]: Sadly, to date I have not found a way to center this. Markdown Tables and markdown resizing will fail and HTML embedding as well.

File: logistic-regression.md (hunk 13 above)
Comment (colinhex) [reply logistic-regression.md:728]: This is equivalent to the partial derivative of the previous expression.

File: logistic-regression.md [hunk 15]
```diff
@@ -0,0 +1,752 @@
...
+x = y = numpy.linspace(-5, 5, 20)
+
+XX, YY = numpy.meshgrid(x, y)
+
+Z = softmax(numpy.dstack((XX, YY)))
+
+fig, ax = plt.subplots(subplot_kw={"projection": "3d"}, figsize=(7,7))
+
+surf1 = ax.plot_surface(XX, YY, Z.T[0].T, cmap=cm.coolwarm, alpha=0.8)
+surf2 = ax.plot_surface(XX, YY, Z.T[1].T, cmap=cm.coolwarm, alpha=0.8)
```
[the remark this answers is not in the document]
Comment (colinhex) [reply logistic-regression.md:410]: In this plot I visualised the softmax-function in two dimensions.

The color is for visual ease, but as it is the z-axis in the plot of the softmax-function, it signifies the probability of one of the two events. The z-axis direction is irrelevant, since it is symmetric over the xy-plane in this case.

File: logistic-regression.md (hunk 10 above)
Comment (colinhex) [reply logistic-regression.md:680]: There is currently no LaTeX algorithm environment package in the repository. The Jupyter-Book documentation references ([here](https://sphinx-proof.readthedocs.io/en/latest/syntax.html#algorithms)) an extra extension for sphinx to enable a algorithm environment, but an inclusion in _config.yml leads to an error:

`Extension error: Could not import extension sphinx_proof (exception: No module named 'sphinx_proof')...[...]`

**Comment by colinhex** [comment]
Hi! I addressed the majority of the comments and added a few additional things. E.g. a cross-entropy figure.

File: logistic-regression.md [hunk 16]
```diff
@@ -0,0 +1,927 @@
...
+ax.text(*(x1 + 0.2), '$x^{(1)}$')
+ax.text(*(x2 + 0.2), '$x^{(2)}$')
+ax.text(*(y + 0.2), '$y$')
+ax.text(*(y_hat + 0.2), '$\widehat{y}$')
+ax.legend()
+plt.axis('off')
+
+ax.view_init(-18, -172)
+
+plt.show()
```
Comment (colinhex) [inline logistic-regression.md:125]: For the vector plot I adhered to the original script while making it more readable. The content of the vectors have no significance here as long as the geometry is clear and were overlapping in the original graphic. Instead I added a legend with algebraic terms.

File: logistic-regression.md [hunk 17]
```diff
@@ -0,0 +1,927 @@
...
+$$
+    \mathbf{y} = \mathbf{w}^T\mathbf{X} + \epsilon \quad \textit{where} \quad \epsilon \sim \mathcal{N}(0, \sigma^2\mathbf{I})
+$$
+
+Adding a normally distributed random variable $\epsilon$ to the equation makes the results random as well, that is logical. The
+only thing to find out is how they are distributed. No problem! The predictions are also normally distributed with variance $\sigma^2$ and mean $\mathbf{w}^T\mathbf{X}$, since for a constant $c$ and 
+normal distribution $F_X$ we have:
+
+$$
+    F_{X + c}(x) = p(X + c \leq x) = p(X \leq x - c) = \int_{-\infty}^{x-c}\frac{1}{\sigma\sqrt{2\pi}}e^{-\frac{1}{2}\left(\frac{(t-\mu)^2}{\sigma}\right)}dt
```
Comment (colinhex) [inline logistic-regression.md:156]: Although it is technically a recap at this point and this should be scribed in a previous chapter, I added an explanation for the shift of the mean with the PDF of the normal distribution.

File: logistic-regression.md [hunk 18]
```diff
@@ -0,0 +1,927 @@
...
+When we speak of the **cross-entropy** between two probability distributions over the same set of events we are talking about
+the average number of information bits we need to classify one of the events. For example, if we have two probability distributions
+$\gamma, \delta$ over the same set of events $\mathcal{X}$, the cross entropy $H$ is defined as:
+
+$$
+    H(\gamma, \delta) = - \sum_{x\in\mathcal{X}}\gamma(x)\ln(\delta(x))
+$$
+
+Let us visualize this for two normal distributions $\gamma(x)\sim\mathcal{N}(s, 1)$ and $\delta(x)\sim\mathcal{N}(0, 1)$. We
+will shift the mean $s$ of $\gamma(x)$ and calculate the cross entropy along the way!
```
Comment (colinhex) [inline logistic-regression.md:478]: I added a visualisation of the cross-entropy of two normally distributed random variables, when the mean of one is shifted.

File: logistic-regression.md [hunk 19]
```diff
@@ -0,0 +1,927 @@
...
+side of this simplification [can be read](https://en.wikipedia.org/wiki/Cross_entropy#Relation_to_maximum_likelihood) as the product of the estimated probabilities for classes $c$ to the
+power of their occurrences in the predictions, which is to say their actual probability. And if we substitute these definitions in the previous equations a seemingly familiar equation....
+    
+$$                                                              
+    \sum_{i=1}^k ln\left(\prod_{c=1}^C \mu^{y_{ic}}_{ic}\right) = \sum_{i=1}^k \sum_{c=1}^C ln(\mu^{y_{ic}}_{ic}) = \sum_{i=1}^k \sum_{c=1}^C y_{ic} ln(\mu_{ic})
+$$
+
+The term $\sum_{c=1}^C y_{ic} ln(\mu_{ic})$ looks suspiciously similar to **cross-entropy**! This means that
+maximizing the likelihood $\ell(w)$ is equivalent to minimizing the cross-entropy, or to spare the headache *minimizing the negative loss-likelihood.* 
+Why is that the case? We are essentially comparing two probability distributions. On the one hand we have the actual distribution of our data, and on
```
Comment (colinhex) [inline logistic-regression.md:546]: And a explanation that fits the visualisation of the cross-entropy is found in the same section.

**Comment by colinhex** [comment]
I addressed your comment regarding gradient descend wrt. gradient steepest descend. However I am finding sources defining terms differently, (gradient descend vs steepest descend vs gradient steepest descend). So [the last commit](https://github.com/swing-research/prbook/pull/6/commits/a4d8c1d78f7fc258d647cb459b2c0c81a4be9ef6) can be taken with a small grain of salt.

File: logistic-regression.md (hunk 14 above)
Comment (vinith2) [reply logistic-regression.md:463]: oh ok, the current setting should be good then.

File: logistic-regression.md (hunk 18 above)
Comment (vinith2) [reply logistic-regression.md:478]: The plot looks good. I think you computed the entropy numerically, For gaussians you can compute it analytically as its just a function of its second order moment. If possible can you add this curve as well?

File: logistic-regression.md (hunk 19 above)
Comment (vinith2) [reply logistic-regression.md:546]: regarding the  statement " suspiciously similar ": The equation is an empirical version of the cross entropy. That is for mean E(x), we write it as  $E(X) = \frac{ \sum x_i}{N}$. Similarly the loss is an empirical version of    cross entropy. Something of the form $E_{x \sim \gamma}(\ln(\delta(x)))$. But i think this is ok for now.

File: logistic-regression.md [hunk 20]
```diff
@@ -0,0 +1,942 @@
...
+
+$$
+    NLL(\mathbf{w}) = -\sum_{i=1}^k \left(y_i ln\left(\frac{1}{1 + e^{w^Tx_i}}\right) + (1-y_i) ln\left(\frac{1}{1 + e^{-w^Tx_i}}\right)\right)
+$$
+
+If we further focus the function for binary cases $y_i \in \{-1, 1\}$, then we can, you are getting the hang of this, *simplify*. 
+First let us consider it with labels $y'_i \in \{0, 1\}$ then for $y'_i = 0$ we have:
+
+$$
+    = -\sum_{i = 1}^k ln\left(\frac{1}{1 + e^{w^Tx_i}}\right)  = \sum_{i = 1}^k ln\left(1 + e^{w^Tx_i}\right)
```
Comment (vinith2) [inline logistic-regression.md:602]: Here you are splitting the sum right? May be add this condition in the sum something like $\sum_{i=0, y_i^{'} = 1}$

File: logistic-regression.md (hunk 10 above)
Comment (vinith2) [reply logistic-regression.md:695]: Oh ok, we will look into this later, thanks for letting me know.  I will keep this issue open for now. So I remember it later.

**Review by vinith2 — COMMENTED** [review]
Hey,
The changes looks good. I have added some comments, but its optional and maybe too much  into the weeds. For now I will keep the PR open.

**Review by vinith2 — APPROVED** [review]
I will approve your changes for now, I will add your name in the headline (for now) so we can evaluate the submission easily.  However there are some minor changes, which we will do it later.
